pulsar-client-cpp/tests/ConsumerConfigurationTest.cc

```diff
@@ -0,0 +1,105 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+#include <pulsar/Client.h>
+#include <gtest/gtest.h>
+
+#include "../lib/Future.h"
+#include "../lib/Utils.h"
+
+using namespace pulsar;
+
+TEST(ConsumerConfigurationTest, testReadCompactPersistentExclusive) {
+    std::string lookupUrl = "pulsar://localhost:8885";
+    std::string topicName = "persistent://prop/unit/ns1/persist-topic";
+    std::string subName = "test-persist-exclusive";
+
+    Result result;
+
+    ConsumerConfiguration config;
+    config.setReadCompacted(true);
+    config.setConsumerType(ConsumerExclusive);
+
+    ClientConfiguration clientConfig;
+    Client client(lookupUrl, clientConfig);
+
+    Consumer consumer;
+    result = client.subscribe(topicName, subName, config, consumer);
+    ASSERT_EQ(ResultOk, result);
+    consumer.close();
+}
+
+TEST(ConsumerConfigurationTest, testReadCompactPersistentFailover) {
+    std::string lookupUrl = "pulsar://localhost:8885";
+    std::string topicName = "persistent://prop/unit/ns1/persist-topic";
+    std::string subName = "test-persist-fail-over";
+
+    Result result;
+
+    ConsumerConfiguration config;
+    config.setReadCompacted(true);
+    config.setConsumerType(ConsumerFailover);
+
+    ClientConfiguration clientConfig;
+    Client client(lookupUrl, clientConfig);
+
+    Consumer consumer;
+    result = client.subscribe(topicName, subName, config, consumer);
+    ASSERT_EQ(ResultOk, result);
+    consumer.close();
+}
+
+TEST(ConsumerConfigurationTest, testReadCompactPersistentShared) {
+    std::string lookupUrl = "pulsar://localhost:8885";
+    std::string topicName = "persistent://prop/unit/ns1/persist-topic";
+    std::string subName = "test-persist-shared";
+
+    Result result;
+
+    ConsumerConfiguration config;
+    config.setReadCompacted(true);
+    config.setConsumerType(ConsumerShared);
+
+    ClientConfiguration clientConfig;
+    Client client(lookupUrl, clientConfig);
+
+    Consumer consumer;
+    result = client.subscribe(topicName, subName, config, consumer);
+    ASSERT_EQ(ResultInvalidConfiguration, result);
+    consumer.close();
+}
+
+TEST(ConsumerConfigurationTest, testReadCompactNonPersistentExclusive) {
+    std::string lookupUrl = "pulsar://localhost:8885";
+    std::string topicName = "non-persistent://prop/unit/ns1/testNonPersistentTopic";
+    std::string subName = "test-non-persist-exclusive";
+
+    Result result;
+
+    ConsumerConfiguration config;
+    config.setReadCompacted(true);
+    config.setConsumerType(ConsumerExclusive);
+
+    ClientConfiguration clientConfig;
+    Client client(lookupUrl, clientConfig);
+
+    Consumer consumer;
+    result = client.subscribe(topicName, subName, config, consumer);
+    ASSERT_EQ(ResultInvalidConfiguration, result);
+    consumer.close();
+}
```

Comment:
@zhaijack As @ivankelly, it would be good to have an end-to-end test here. Since the broker is running when the tests are executed, it's possible to make HTTP calls to trigger the compaction and verify the client indeed receives only the compacted data. 

It might be easier from the Python unit tests (at least in terms of making the rest calls).

Reply:
Thanks, will try to use python